### PR TITLE
Add a new endpoint to generate PDFs with CMYK text 

### PR DESCRIPTION
**Pull request by quis**
_This is an initial stab at getting this working, which won’t yet be used anywhere._

_Adapted from https://github.com/riedel/pdf2indexedCMYK._

***

Currently the PDFs that this app produces have their colours specified in RGB, which is intended for screen display, not print.

The problem with trying to print RGB text, or text that has been naïvely converted to CMYK, is that the edges of the letterforms are fuzzy because the multiple ink colours don’t exactly line up.

This PR lets us produce PDFs with text that is 100% K ink, rather than a mixture of C, M, Y and K. This results in sharp letterforms when printing.

It does this by adding a new endpoint (`/print.pdf`) which:

- generates the RGB PDF as before (using Weasyprint)
- pipes the RGB PDF into PDFTK, a command line programme that can decompress PDF data into PDF code
- goes through each line of the PDF code looking for colours to replace
- pipes the replaced PDF code into GhostScript, which writes it back into compressed PDF data

***

image

### CODE REVIEW

**Comment by quis** [comment]
# Use print version of the letter template renderer

This is a special version of our letter template which remove the barcodes and other things that we don’t want to send to DVLA.

Used only on the printable version, since we still want to show a preview with these things in.

image

**Comment by venusbb** [comment]
The function works on my local docker. I have noticed:-
- A simple page conversion takes a second
- print.pdf do not have the 2D bar codes that preview.pdf has
- The yellow colour that highlight the addresses appear different on screen.

File: app/preview.py [hunk 1]
```diff
@@ -62,6 +38,13 @@ def png_from_pdf(pdf_endpoint, page_number):
     }
 
 
+def get_logo_filename(dvla_org_id):
+    try:
+        return current_app.config['LOGO_FILENAMES'][dvla_org_id]
+    except KeyError:
+        abort(400)
+
```
Comment (venusbb) [inline app/preview.py]: Write a unit test for this function.

**Comment by venusbb** [comment]
How do we know if this library is working as expected? Can we build some unit tests around it?

**Comment by venusbb** [comment]
Are there any situation that it won't work?

**Comment by venusbb** [comment]
In the example black is shown. Have we tested other colours?

**Comment by venusbb** [comment]
What is the CMYK profile we use?

File: tests/test_preview.py [hunk 2]
```diff
@@ -166,3 +189,10 @@ def test_page_count(
     )
     assert response.status_code == 200
     assert json.loads(response.get_data(as_text=True)) == {'count': expected_pages}
+
+
+def test_print_letter_returns_200(print_letter_template):
+    resp = print_letter_template()
+
+    assert resp.status_code == 200
```
[the remark this answers is not in the document]
Comment (quis) [reply tests/test_preview.py]: What do you mean by local file?

File: app/preview.py (hunk 1 above)
Comment (quis) [reply app/preview.py]: Test added in 3ff0394

**Comment by quis** [comment]
This is only tested with black and yellow at the moment. For now, black is the only colour we should have in real letters that we’re printing (ignoring logos).

No need for any colour profile – we just want them to print the exact amount of ink we’re specifying.

File: tests/test_preview.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (venusbb) [reply tests/test_preview.py]: logo file - get_logo_filename

File: app/preview.py (hunk 1 above)
Comment (venusbb) [reply app/preview.py]: Can you add a TODO about the logo issue you mentioned yesterday?

File: app/preview.py [hunk 3]
```diff
@@ -120,3 +98,47 @@ def view_letter_template(filetype):
...
         raise e
+
+
+@preview_blueprint.route("/print.pdf", methods=['POST'])
+@auth.login_required
+def print_letter_template():
+    """
+    POST /print.pdf with the following json blob
+    {
+        "letter_contact_block": "contact block for service, if any",
```
[the remark this answers is not in the document]
Comment (quis) [reply app/preview.py]: Added in 4c20eb4

File: app/preview.py [hunk 4]
```diff
@@ -120,3 +98,47 @@ def view_letter_template(filetype):
...
+        json = get_and_validate_json_from_request(request, preview_schema)
+        logo_file_name = get_logo_filename(json['dvla_org_id'])
+
+        template = LetterPrintTemplate(
+            json['template'],
+            values=json['values'] or None,
+            contact_block=json['letter_contact_block'],
+            # we get the images of our local server to keep network topography clean,
+            # which is just http://localhost:6013
+            admin_base_url='http://localhost:6013',
```
[the remark this answers is not in the document]
Comment (quis) [reply app/preview.py]: It doesn’t change from one environment to another.

File: app/transformation.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import sys
+import yaml
+import os
+import subprocess
+import numpy as np
+import re
+
+
+class PDFTKError(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (quis) [reply app/transformation.py]: `pass` doesn’t mean the exception is suppressed, it’s just a way of defining a class or function that does’t have any methods, attributes or a body.

> pass is a null operation — when it is executed, nothing happens

– https://docs.python.org/3.2/reference/simple_stmts.html#the-pass-statement

File: app/transformation.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        ] = 'C {}, M {}, Y {}, K {}'.format(*cmyk_tuple)
+
+    def get_cmyk_color_from_match(self, match):
+        rgb_tuple = tuple(float(match.group(i)) for i in (2, 3, 4))
+        cmyk_tuple = self.get_closest_cmyk_color(rgb_tuple)
+        self.log_replacement(rgb_tuple, cmyk_tuple)
+        return '{} {} {} {}'.format(*cmyk_tuple).encode('UTF-8')
+
+
+class PDFData:
```
[the remark this answers is not in the document]
Comment (quis) [reply app/transformation.py]: I think 140 lines for a file is fine.

File: app/preview.py (hunk 1 above)
Comment (quis) [reply app/preview.py]: Added a story: https://www.pivotaltracker.com/story/show/153137328

File: tests/test_preview.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (quis) [reply tests/test_preview.py]: Test added in 5724700

File: app/preview.py [hunk 7]
```diff
@@ -120,3 +98,47 @@ def view_letter_template(filetype):
...
+        )
+        html = HTML(string=str(template))
+        pdf = render_pdf(html)
+
+        with PDFData(pdf.get_data()) as pdf_data:
+            for line in pdf_data.read():
+                pdf_data.write(color_mapping(line))
+
+        return send_file(
+            BytesIO(pdf_data.result),
```
[the remark this answers is not in the document]
Comment (quis) [reply app/preview.py]: Good catch. Added in 9e62c93.

**Pull request merged**